src/server/lock_manager/waiter_manager.rs

```diff
@@ -232,13 +242,16 @@ impl WaiterManager {
         }
     }
 
-    fn handle_wait_for(&mut self, handle: &Handle, waiter: Waiter) {
+    fn handle_wait_for(&mut self, handle: &Handle, waiter: Waiter, mut timeout: u64) {
         let lock = waiter.lock;
         let start_ts = waiter.start_ts;
 
         if self.wait_table.borrow_mut().add_waiter(lock.ts, waiter) {
             let wait_table = Rc::clone(&self.wait_table);
-            let when = Instant::now() + Duration::from_millis(self.wait_for_lock_timeout);
+            if timeout == 0 {
```

Comment:
May need to make sure the `timeout` is not longer than `wait_for_lock_timeout` for safety.
`if timout == 0 || timeout > self.wait_for_lock_timeout`

Reply:
Then, how can we wait longer than the default lock timeout?

Reply:
We don't need to and should not wait too long, otherwise, the deadlock error may not be detected in time.

Reply:
InnoDB has default lock timeout of 50 seconds. It is far longer than our current default timeout. So to my understanding the limitation is due to the deadlock detection mechanism which doesn't ensure deadlocks can be detected.

Reply:
We depend on retry mechanism on tidb-server side to make sure the WaitFor information is not lost due to region leader change and deadlock detector leader change.

Reply:
The retry mechanism on tidb-server is necessary. If the region leader is changed, the commit request won't send to the tikv which has waiters waiting for the lock. These waiters won't be waked up.